Fix #110: Intrinsics and extrinsics

astojilj opened the pull request:
Fix #110: Intrinsics and extrinsics
    
Principal point, distortion coefficients, depth to video transformation.
Formulas for deprojection, transform and projection.
TODO. shader example code.    
Fix #110: Accessing Camera Calibrations

astojilj committed:
Fix #110: Intrinsics and extrinsics

principal, distortion coefficients, depth to video.
Formulas for deprojection, transform and projection.

Fix #110: Accessing Camera Calibrations

astojilj commented:
@huningxin, @robman, @bbernhar, @fluffy, PTAL.
Thanks.

astojilj commented:
@huningxin, I know you are busy. Some feedback would be appreciated as I'm starting work on writing formulas for deprojection etc... Thanks.

huningxin commented:
@astojilj , Thanks for your PR! It's my bad that I didn't comment here. 

But I understand that we still have some opens in #110 and #135.  

My suggestion would be only including the changes with consensus in this PR and leave the opens in following PRs. WDYT?

astojilj commented:
@huningxin, @robman, @anssiko 
>But I understand that we still have some opens in #110 and #135.

#135:
I don't see there is anything open in #135. All the comments and reporter confirm that this needs to be fixed and the PR here is a fix for it. in [comment here](https://github.com/w3c/mediacapture-depth/issues/135#issuecomment-260585403) @robman expresses agreement with my approach coded here in this PR.

#110,
The only open there is @robman's [proposal](https://github.com/w3c/mediacapture-depth/issues/110#issuecomment-260587927) about which part of this PR should be v1 and which should be v2 scope. You also raised the [question](https://github.com/w3c/mediacapture-depth/issues/110#issuecomment-260875248) here about why to attempt to provide incomplete data only.

I don't think that PR should matter about v1/v2 scope split. Or in PR only we should put content that all the participants in issue page agree should be v1  and delete all v2 scope content?

huningxin commented:
Then how about separate this PR into one for #135 and another one for #110. It would be more clear to me. Thanks!

astojilj commented:
Yes, let me do so. Thanks.

astojilj commented:
PR for #135 is split to https://github.com/w3c/mediacapture-depth/pull/146.

anssiko commented:
Updated the title, since #135 was already fixed with https://github.com/w3c/mediacapture-depth/commit/e50bc08af1bc6e520d87dd66d6842191b322ac6c

astojilj commented:
Updated. PTAL. 
Travis CI fails on acceptable "Warning: trimming empty <span>" only.

astojilj commented:
After peer review with Anssi:
- need to generalize distortion model and coefficients (probably remove the model name)
- provide generic formula for applying distortion coefficients.

astojilj commented:
Updated the pull request by:
1) Removed device specific distortion models
2) added full formulas for depth deprojection, translating from depth to color coordinates and color projection with the optional distortion handling formular on both deptojection and projection.
3) transformation matrix was modeled using translation and quaternion for rotation - replaced them by simple 16 element array in the format suitable to be directly used in matrix x vector multiplication and with WebGL's uniformMatrix4fv.

Example code for color deprojection and projection I'll add in separate request - don't want to grow this one further.

anssiko commented on index.src.html:
```diff
@@ -345,15 +345,232 @@
+        </p>
+        <p>
+          A <a>depth map</a> has an associated <dfn>transformation from depth
+          to video</dfn>, which is <a>DepthToVideoTransform</a> dictionary. It
```
`which is a <dfn>transformation matrix</dfn> represented by a <a>DepthToVideoTransform</a> dictionary`

astojilj replied:
Thanks. Done.

anssiko commented on index.src.html:
```diff
@@ -345,15 +345,232 @@
+          both of the coefficients specified. The most common scenario is that
+          the depth track has deprojection coefficients or that the color video
+          track has projection coefficients.
+        </p>
```
L390 onwards could be in its own section, for example (adjust titles so they make sense):

* Coordinate transformation from depth to video
  * Deproject depth map values to points in depth 3D space
  * Transform 3D points from depth camera 3D space to color camera 3D space
  * Project from color camera 3D space to color frame 2D pixels
* Conversion from the depth map values to grayscale
  * (this would contain the current _convert the depth map value to grayscale_ and _rules to convert using range linear_ algorithms)

astojilj replied:
_Coordinate transformation from depth to video_ is separated now.
_Conversion from the depth map values to grayscale_ I'm doing in separate patch as it requires content changes.

anssiko commented on index.src.html:
```diff
@@ -345,15 +345,232 @@
+          track has projection coefficients.
+        </p>
+        <p>
+          With this case in mind, the algorithm for mapping depth pixels to
```
`The <dfn>algorithm to map depth pixels to color pixels</dfn> is as follows:`

astojilj replied:
done.

anssiko commented on index.src.html:
```diff
@@ -345,15 +345,232 @@
+          color pixels is as follows:
+        </p>
+        <ol>
+          <li>Deproject <a>depth map</a> values to points in depth 3D space.
```
`When the user agent is to <dfn>deproject depth map values</dfn> to points in depth 3D space, it must run the following algorithm:`

astojilj replied:
Application or user agent. rephrased.

anssiko commented on index.src.html:
```diff
@@ -345,15 +345,232 @@
+            </p>
+            <p>
+              Let <var>fx</var> and <var>fy</var> be <a>depth map</a>'s
+              horizontal and vertical <a>focal length</a>s.
```
`<a>horizontal focal length</a> and <a>vertical focal length</a> respectively.

astojilj replied:
done.

anssiko commented on index.src.html:
```diff
@@ -345,15 +345,232 @@
+            </p>
+            <ul>
+              <li>If <a>depth map</a>'s <a>deprojection distortion
+              coefficients</a> are not available,
```
Must define what "not available" means, `DistortionCoefficients` is set to false?

astojilj replied:
i use omitted for undefined and defined instead of available now.

in JS, "defined" would be  `if(track.getSettings().deprojectionDistortionCoefficients)   ` - added this note as inline example.

anssiko commented on index.src.html:
```diff
@@ -345,15 +345,232 @@
+              </li>
+              <li>If <a>depth map</a>'s <a>deprojection distortion
+              coefficients</a> <a>k1</a>, <a>k2</a>, <a>k3</a>, <a>p1</a> and
+              <a>p2</a> are available, with a note that some of those could be
```
s/are available/are not omitted/

astojilj replied:
defined?

anssiko commented on index.src.html:
```diff
@@ -345,15 +345,232 @@
+              <a>p2</a> are available, with a note that some of those could be
+              zero,
+                <p>
+                  3D coordinates (Xd, Yd, Zd) in depth camera space are
```
`When the user agent is to <dfn>calculate 3D coordinates in depth camera space</dfn> (Xd, Yd, Zd), it must run the following algorithm:`

anssiko commented on index.src.html:
```diff
@@ -345,15 +345,232 @@
+          <li>Transform 3D points from depth camera 3D space to color camera 3D
+          space.
+            <p>
+              The 3D point (Xd, Yd, Zd) calculated in previous step is in depth
```
Link to the previous section rather than say "previous step":

To cross-link easily, you can say `<a>foo</a>` to link to `<dfn>foo</dfn>` globally in the document.

astojilj replied:
done.

anssiko commented on index.src.html:
```diff
@@ -345,15 +345,232 @@
+            <p>
+              For projection we use color track's <a>projection distortion
+              coefficients</a> <a>k1</a>, <a>k2</a>, <a>k3</a>, <a>p1</a> and
+              <a>p2</a> if available. The color track that corresponds to the
```
s/if available/if not omitted/

What happens if they are omitted? Define failure path.

astojilj replied:
It is defined under - rewrote this part so there is no confusion. "available" should probably be "defined".

anssiko commented on index.src.html:
```diff
@@ -345,15 +345,232 @@
+                </p>
+              </li>
+              <li>If color track's <a>projection distortion coefficients</a>
+              are not available
```
are omitted

astojilj replied:
done.

anssiko commented:
Thanks! I added some review comments.

In addition, you'll need to patch the build.js as follows to make MathJax happy with the complex formulas:

https://gist.github.com/anssiko/fd03fb67eb7b29dd1abecbac0ec43536

astojilj reviewed (commented):
PR updated.
Thanks.

anssiko commented:
Regarding checking the presence of dictionary members, we should actually check if the member is present, and refer to https://heycam.github.io/webidl/#dfn-present like this:

`"If the <a>foobar</a> member is <a>present</a> in <a>baz dictionary</a>"`

(Additionally, add appropriate `<dfn>present</dfn>` to Dependencies.

Related issue: https://github.com/heycam/webidl/issues/124

astojilj commented:
<a>present</a> and <a>not present</a> in <a href="#mediatracksettings-dictionary">MediaTrackSettings</a> dictionary in updated PR.

anssiko commented on index.src.html:
```diff
@@ -326,15 +331,51 @@
+          to translate position in depth camera 3D coordinate system to RGB
+          video stream's camera (identified by <dfn>videoDeviceId</dfn>) 3D
+          coordinate system. After projecting depth 2D pixel coordinates to 3D
+          space, we use this to translate depth camera 3D space coordinates to
```
we use this projection?

astojilj replied:
"this matrix to transform"

anssiko commented on index.src.html:
```diff
@@ -326,15 +331,51 @@
+          Both depth and color cameras usually introduce significant distortion
+          caused by the camera and lens. While in some cases, the effects are
+          not noticeable, these distortions cause errors in image analysis. To
+          map <a>depth map</a> pixel values to a corresponding color video
```
to corresponding

astojilj replied:
done.

anssiko commented on index.src.html:
```diff
@@ -326,15 +331,51 @@
+          distortion coefficients</dfn>.
+        </p>
+        <p>
+          Deprojection distortion coefficients are used for compensating camera
```
`<a>Deprojection distortion coefficients</a>`

astojilj replied:
done

anssiko commented on index.src.html:
```diff
@@ -326,15 +331,51 @@
+        <p>
+          Deprojection distortion coefficients are used for compensating camera
+          distortion when deprojecting 2D pixel coordinates to 3D space
+          coordinates. Projection distortion coefficients are used in the
```
`<a>Projection distortion coefficients</a>`

astojilj replied:
done

anssiko commented on index.src.html:
```diff
@@ -326,15 +331,51 @@
+          coordinates. Projection distortion coefficients are used in the
+          opposite case, when projecting camera 3D space points to pixels. One
+          track doesn't have both of the coefficients specified. The most
+          common scenario is that the depth track has deprojection coefficients
```
`<a>deprojection distortion coefficients</a>`

astojilj replied:
done

anssiko commented on index.src.html:
```diff
@@ -326,15 +331,51 @@
+          opposite case, when projecting camera 3D space points to pixels. One
+          track doesn't have both of the coefficients specified. The most
+          common scenario is that the depth track has deprojection coefficients
+          or that the color video track has projection coefficients. For the
```
`<a>projection distortion coefficients</a>`

astojilj replied:
done

anssiko commented on index.src.html:
```diff
@@ -394,6 +435,233 @@
+        Synchronizing depth and color video rendering
+      </h2>
+      <p>
+        When rendering, we want to position a color from color video frame to
```
a color value?

astojilj replied:
done

anssiko commented on index.src.html:
```diff
@@ -394,6 +435,233 @@
+        coefficients</a> to compensate camera distortion when deprojecting 2D
+        pixel coordinates to 3D space coordinates and <a>projection distortion
+        coefficients</a> in the opposite case, when projecting camera 3D space
+        points to pixels. The algorithm here intents to cover all of the cases.
```
`The <a>algorithm to map...</a> and <a>algorithm to deproject...</a> are defined in this specifications.`

astojilj replied:
redundant sentence, removed.

anssiko commented on index.src.html:
```diff
@@ -394,6 +435,233 @@
+        pixel coordinates to 3D space coordinates and <a>projection distortion
+        coefficients</a> in the opposite case, when projecting camera 3D space
+        points to pixels. The algorithm here intents to cover all of the cases.
+        Most commonly, either the depth track has deprojection or the color
```
We already say above that "The most common scenario is that the depth track has deprojection coefficients or that the color video track has projection coefficients" so we do not need to repeat.

astojilj replied:
done.

anssiko commented on index.src.html:
```diff
@@ -394,6 +435,233 @@
+          <dfn>Deproject</dfn> to depth 3D space
+        </h2>
+        <p>
+          To <a>deproject</a> <a>depth map value</a> to point in depth camera
```
`The <dfn>algorithm to deproject depth map value</dfn> to point in depth camera is as follows:`

anssiko replied:
```
To <dfn>deproject</dfn> a depth map value to point in depth camera 3D space, the user agent must run the <a>algorithm to deproject depth map value</a>.

(Remove dfns from headers.)

The <dfn>algorithm to deproject depth map value</dfn> to point in depth camera is as follows:
```

anssiko replied:
Alternative, more concise way to say this (if you won't need to link to the algorithm directly elsewhere in the spec):
`To <dfn>deproject</dfn> a depth map value to point in depth camera 3D space, the user agent must run the following steps:`

astojilj replied:
done.

anssiko commented on index.src.html:
```diff
@@ -394,6 +435,233 @@
+          To <a>deproject</a> <a>depth map value</a> to point in depth camera
+          3D space, application or user agent must run following algorithm:
+        </p>
+        <p>
```
Make this and other algorithms numbered lists, nested at ifs:

```
<ol> 
  <li>...</li> 
  <li>If ...
    <ol> 
      <li>inner ...</li> 
      <li>...</li> 
    </ol>
  </li>
  <li>...</li> 
</ol> 
```

astojilj replied:
ok, agreed to leave it as it is.

anssiko commented on index.src.html:
```diff
@@ -394,6 +435,233 @@
+          <dfn>Transform</dfn> from depth to color 3D space
+        </h2>
+        <p>
+          The 3D point (Xd, Yd, Zd), that is a result of <a>project</a> depth
```
`To <dfn>transform</dfn> xxxxxx, the user agent must run the <a>algorithm to transform yyyy</a>.`

`The <dfn>algorithm to transform yyyy</dfn> is as follows:`

(Remove dfn from the header.)

This allows us to decouple the "transform" concept from its algorithm, so we can link to them separately as needed elsewhere in the spec.

anssiko replied:
(See related comment above for more concise way to spec this.)

anssiko commented on index.src.html:
```diff
@@ -394,6 +435,233 @@
+          <a>transformation from depth to video</a> matrix by the (Xd, Yd, Zd)
+          3D point vector.
+        </p>
+        <p>
```
Make this a numbered list.

astojilj replied:
acknowledged. ignored as this goes to example section.

anssiko commented on index.src.html:
```diff
@@ -394,6 +435,233 @@
+          <dfn>Project</dfn> from color 3D to pixel
+        </h2>
+        <p>
+          With this step, we conclude mapping of a <a>depth map value</a>
```
Here, use the same style as above:

`To <dfn>project</dfn> ...`

anssiko commented on index.src.html:
```diff
@@ -394,6 +435,233 @@
+          "#mediatracksettings-dictionary">MediaTrackSettings</a>.
+        </p>
+        <p>
+          Let `fx_c` and `fy_c` be color track's horizontal and vertical
```
`<a>color stream track</a>'s <a>horizontal focal length</a> and <a>vertical focal length</a> respectively.`

anssiko commented on index.src.html:
```diff
@@ -394,6 +435,233 @@
+          With this step, we conclude mapping of a <a>depth map value</a>
+          coordinate to a pixel value in color video frame.
+        </p>
+        <p>
```
It sounds like this para is informative content? If so, use:
`<p class="note">`

astojilj replied:
removed as not very informative.

anssiko commented on index.src.html:
```diff
@@ -394,6 +435,233 @@
+          <a>getSettings()</a> to access <a href=
+          "#mediatracksettings-dictionary">MediaTrackSettings</a>.
+        </p>
+        <p>
```
`<p>To <dfn>project</dfn> xxxxx, the user agent must run the following steps:</p>`

And make the steps a numbered list.

anssiko commented:
Discussed with @astojilj and concluded "Synchronizing depth and color video rendering" section will be made non-normative guidance to web developers. Thus some of my earlier comments are not valid, since we should not use RFC terms in an informative section.

astojilj commented:
PR updated. Section moved just before examples, Example vertex shader code added and links to it added around.
@ds-hwang Kudos for the [code](https://github.com/IntelRealSense/librealsense/pull/168/files#diff-eb62c5bc83e467cd92828ec977b8a77bR439) - I used it many times last week.

astojilj commented:
@anssiko PTAL

anssiko commented on build.js:
```diff
@@ -62,15 +62,19 @@ mjAPI.typeset({
   // remove the MathJax @font-face CSS at-rules
   // var re4 = /@font-face {font-family: MJX(.*?)}\n/gi;
 
+  var re4 = /.mjx-char {display: block; white-space: pre}\n/gi;
```
re4 seems redundant, is not used in L72.

anssiko commented on index.src.html:
```diff
@@ -660,8 +883,104 @@
+            </dd>
+            <dt>
+              The <dfn><code>deprojectionDistortionCoefficients</code></dfn>
+              dictionary member represents the media track's (usually the
```
Ambiguos. Suggestion: "represents depth map's distortion...".

anssiko commented on index.src.html:
```diff
@@ -660,8 +883,104 @@
+        <div data-dfn-for="DistortionCoefficients">
+          <dl class="dictionary-members">
+            <dt>
+              The <dfn>k1</dfn>, <dfn>k2</dfn>, <dfn>p1</dfn>, <dfn>p2</dfn>
```
Without further context unclear what the semantics of the coefficients are.

astojilj replied:
Note added:

> Radial distortion coefficients and tangential distortion coefficients are used when there's need to deproject depth value to 3D space or to project 3D value to 2D video frame coordinates.
> See algorithm to map depth pixels to color pixels and Brown-Conrady distortion model implementation 3D point cloud rendering example GLSL shader.
> 
> 

Don't have a good reference to Brown-Conrady model . Wikipedia [link](https://en.wikipedia.org/wiki/Distortion_(optics)#Software_correction)? 

Also, I feel we will need to add examples for the cases listed by @bbernhar in #110 but that would need to come later.

anssiko commented on index.src.html:
```diff
@@ -660,8 +883,104 @@
+            </dd>
+            <dt>
+              The <dfn><code>videoDeviceId</code></dfn> dictionary member
+              represents the deviceId of video camera the depth stream could be
```
must be syncronized with

anssiko commented on index.src.html:
```diff
@@ -660,8 +883,104 @@
+            <dt>
+              The <dfn><code>videoDeviceId</code></dfn> dictionary member
+              represents the deviceId of video camera the depth stream could be
+              synchronized with. The deviceId can be used as a constraint in
```
Make "The deviceId ..." a note.

anssiko commented on index.src.html:
```diff
@@ -778,6 +1097,15 @@
           </p>
+          <p class="note">
+            This approach is simple to use but comes with the following
+            caveats: it might not be available and the resolutions of two
```
might not be supported by the implementation

anssiko commented on index.src.html:
```diff
@@ -778,6 +1097,15 @@
+          <p class="note">
+            This approach is simple to use but comes with the following
+            caveats: it might not be available and the resolutions of two
+            tracks are intended to be the same. The alternative approach is
```
same that can require downscaling and degrade quality.

astojilj replied:
It is now like you suggested:
I have omitted these two reasons - they could be even more important that the quality but maybe they shouldn't be here:
1. bandwidth and performance issue
The API is not good because it, by the design to be used on canvas2D, it implies significant performance hit: 
- to get downsampled content for use on canvas, we'll need to get the texture back from a) GPU to CPU (if sync and dowscalling happens there) or do the b) sync of two on CPU.
either way, sub optimal implementation.

2. As @bbernhar argumented in #110, it is also about the lost of precision  required for recognition use cases.

IMHO, we need to try different API approach for canvas2D.

anssiko commented on index.src.html:
```diff
@@ -778,6 +1097,15 @@
+            tracks are intended to be the same. The alternative approach is
+            that a web developer implements, following the <a>algorithm to map
+            depth pixels to color pixels</a> and the corresponding <a href=
+            "https://www.w3.org/wiki/Media_Capture_Depth_Stream_Extension#Examples">
```
Link to the example section, not to the wiki.

Suggested rewording: 

The alternative approach is that a web developer implements the algorithm to map depth pixels to color pixels in WebGL vertex shader.

astojilj replied:
Thanks, removed the part "in WebGL vertex shader." as for recognition use cases, CPU implementation could be valid approach too (e.g. when processing 1080p color frame developer is interested in depth value for the specific pixel and doesn't necessary require full color frame processing).

anssiko commented on index.src.html:
```diff
@@ -842,16 +1170,71 @@
-          conversion from the depth value <var>d</var> to a quantized
-          d<sub>8bit</sub> and back to an approximation of the depth value
-          <var>d</var>.
+          The <code>near</code> and <code>far</code> constrainable properties,
```
depthNear and depthFar

astojilj commented:
PR Updated
@anssiko PTAL.

Open issue:

Added this note:

> Radial distortion coefficients and tangential distortion coefficients are used when there's need to deproject depth value to 3D space or to project 3D value to 2D video frame coordinates.
> See algorithm to map depth pixels to color pixels and Brown-Conrady distortion model implementation 3D point cloud rendering example GLSL shader.
> 
> 

Don't have a good reference to Brown-Conrady model . Wikipedia [link](https://en.wikipedia.org/wiki/Distortion_(optics)#Software_correction)?

anssiko commented:
Thanks @astojilj! I'll land this now and do some editorial cleanup afterwards.
